ruby/src/IceRuby/Util.cpp

```diff
@@ -677,41 +545,103 @@ IceRuby::createArray(long sz)
     return arr;
 }
 
-VALUE
-IceRuby::convertLocalException(std::exception_ptr eptr)
+namespace
 {
-    //
-    // We cannot throw a C++ exception or raise a Ruby exception. If an error
-    // occurs while we are converting the exception, we do our best to return
-    // an appropriate Ruby exception.
-    //
-    try
-    {
-        rethrow_exception(eptr);
-    }
-    catch (const Ice::LocalException& ex)
+    template<size_t N>
+    VALUE createRubyException(const char* typeId, std::array<VALUE, N> args, bool fallbackToLocalException = false)
     {
+        string className = string{typeId}.substr(2);
+        VALUE rubyClass = T_NIL;
         try
         {
-            string name = string{ex.ice_id()}.substr(2);
-            volatile VALUE cls = callRuby(rb_path2class, name.c_str());
-            if (NIL_P(cls))
+            // callRuby throws a RubyException if rb_path2class fails.
+            rubyClass = callRuby(rb_path2class, className.c_str());
+        }
+        catch (const RubyException&)
+        {
+            if (fallbackToLocalException)
+            {
+                rubyClass = callRuby(rb_path2class, "Ice::LocalException");
+            }
+            else
             {
-                throw RubyException(rb_eRuntimeError, "exception class `%s' not found", name.c_str());
+                throw;
             }
-            volatile VALUE result = callRuby(rb_class_new_instance, 0, reinterpret_cast<VALUE*>(0), cls);
-            setExceptionMembers(eptr, result);
-            return result;
         }
-        catch (const RubyException& e)
+        return callRuby(rb_class_new_instance, N, args.data(), rubyClass);
+    }
+}
+
+VALUE
+IceRuby::convertException(std::exception_ptr eptr)
+{
+    const char* const localExceptionTypeId = "::Ice::LocalException";
+
+    // We cannot throw a C++ exception or raise a Ruby exception. If an error occurs while we are converting the
+    // exception, we do our best to _return_ an appropriate Ruby exception.
+    try
+    {
+        try
+        {
+            rethrow_exception(eptr);
+        }
+        // First handle exceptions with extra fields we want to provide to Ruby users.
+        catch (const Ice::RequestFailedException& ex)
+        {
+            std::array args = {
```

Comment:
The previous setup prevents the GC of the arguments by using `volatile VALUE`, my understanding is that with this code once we call Ruby if the GC runs there is nothing to prevent it from collecting these arg VALUEs.

Reply:
I am not following here. At which line can we run into trouble with this PR?

We transmit these args VALUE to createRubyException, so they remain on the stack.
And in createRubyException, we use args on the last line (callRuby). So the optimizer can't get rid of them earlier.

Reply:
Once we call Ruby GC is free to run, and because the VALUE is not visible in the Stack it can be GCd. The Ruby GC would not check the array contents as roots for the mark and swipe algorithm.

Reply:
> The Ruby GC would not check the array contents as roots for the mark and swipe algorithm.

Why not? The array contents is on the stack as well.
https://en.cppreference.com/w/cpp/container/array

Reply:
I don't think GC will collect them prematurely there. I don't think there's risk of a compiler somehow optimizing them away here.

Reply:
> Why not?

They only consider root objects the VALUE types on the stack, and std::array<VALUE> is not of type VALUE and would not be considered a root object. Here is not about optimizations.

Reply:
Because the type of the frame on the stack is not VALUE. My understanding is that the type of the frame in the stack is std::arrary<VALUE, 4>. 

The algorithm looks at frames on the STACK if they are of type VALUE they are considered roots, if they are any other type they are not.

Reply:
My understanding is the bytes on the stack are identical. See:
https://coliru.stacked-crooked.com/a/d02cf7782283d5fd

How does Ruby GC tell the difference, if the bytes are the same?